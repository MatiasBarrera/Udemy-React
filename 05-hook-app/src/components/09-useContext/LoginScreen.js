import React, { useContext } from 'react'
import { UserContext } from './UserContext'

export const LoginScreen = () => {
  const { setUser } = useContext(UserContext)

  const handlerSetUser = () => {
    const user = {
      id: 123,
      userName: 'Barrera',
      email: 'matias.barrera95@gmail.com'
    }

    setUser(user)
  }
  return (
    <div>
      <h1>Login</h1>
      <hr />
      <button className="btn btn-primary" onClick={handlerSetUser}>
        Login
      </button>
    </div>
  )
}
