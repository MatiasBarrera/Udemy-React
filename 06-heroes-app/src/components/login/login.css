.login-container {
  align-items: center;
  background-color: #fff;
  display: flex;
  flex-direction: column;
  height: 100vh;
  padding-top: 2rem;
  width: 100vw;
}

.login-card-container {
  background-color: #334756;
  border-radius: 10px;
  width: 500px;
  height: 500px;
}

.login-card-header h1 {
  text-align: center;
  font-size: 3em;
  padding: 0.5em 0px;
  color: #fff;
}

.login-card-body {
  display: flex;
  justify-content: center;
  padding: 2em 0px;
}

.login-card-body button {
  border: none;
  padding: 0.5rem 3rem;
  border-radius: 10px;
  background-color: #ff4c29;
  font-size: 1em;
  font-weight: bold;
  cursor: pointer;
}

.login-card-body button:hover {
  background-color: #fe8f8f;
  color: #fff;
  transition: all 0.4s;
}
