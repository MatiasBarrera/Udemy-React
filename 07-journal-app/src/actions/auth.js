import {
  getAuth,
  signInWithPopup,
  createUserWithEmailAndPassword,
  updateProfile,
  signInWithEmailAndPassword,
  signOut
} from '@firebase/auth'
import Swal from 'sweetalert2'
import { googleAuthProvider } from '../firebase/firebase-config'
import { types } from '../types/types'
import { FinishLoading, StartLoading } from './ui'

export const startLoginEmailPassword = (email, password) => {
  return dispatch => {
    const auth = getAuth()

    dispatch(StartLoading())
    signInWithEmailAndPassword(auth, email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName))
        dispatch(FinishLoading())
      })
      .catch(e => {
        Swal.fire({
          icon: 'error',
          title: 'Error',
          text: 'There is no user record corresponding to this identifier. The user may have been deleted.'
        })
        dispatch(FinishLoading())
      })
  }
}

export const startRegisterWithEmailPasswordName = (name, email, password) => {
  return dispatch => {
    const auth = getAuth()
    createUserWithEmailAndPassword(auth, email, password)
      .then(async ({ user }) => {
        await updateProfile(user, { displayName: name })

        dispatch(login(user.uid, user.displayName))
      })
      .catch(e => {
        console.log(e)
      })
  }
}

export const startGoogleLogin = () => {
  return dispatch => {
    const auth = getAuth()
    signInWithPopup(auth, googleAuthProvider).then(({ user }) => {
      dispatch(login(user.uid, user.displayName))
    })
  }
}

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    name: displayName
  }
})

export const startLogout = () => {
  return async dispatch => {
    const auth = getAuth()
    await signOut(auth)
    dispatch(logout())
  }
}

export const logout = () => ({
  type: types.logout
})
