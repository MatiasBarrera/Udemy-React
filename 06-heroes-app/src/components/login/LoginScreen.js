/* eslint-disable no-unused-vars */
import React, { useContext } from 'react'
import { AuthContext } from '../../auth/AuthContext'
import { useHistory } from 'react-router-dom'
import { types } from '../../types/types'
import './login.css'

export const LoginScreen = () => {
  const history = useHistory()
  const { dispatch } = useContext(AuthContext)

  const handleLogin = () => {
    const lastPath = localStorage.getItem('lastPath') || '/'
    dispatch({
      type: types.login,
      payload: {
        name: 'Matias'
      }
    })

    history.replace(lastPath)
  }

  return (
    <div className="login-container">
      <div className="login-card-container">
        <div className="login-card-header">
          <h1>Welcome</h1>
        </div>
        <div className="login-card-body">
          <button onClick={handleLogin}>Login</button>
        </div>
      </div>
    </div>
  )
}
