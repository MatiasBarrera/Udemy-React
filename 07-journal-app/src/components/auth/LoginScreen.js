import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth'
import googleLogo from '../../assets/google.logo.svg'
import { useForm } from '../../hooks/useForm'

export const LoginScreen = () => {
  const dispatch = useDispatch()
  const { loading } = useSelector(state => state.ui)

  const [formValues, handleInputChange] = useForm({
    email: 'matias@gmail.com',
    password: '123456'
  })

  const handleLogin = e => {
    e.preventDefault()
    dispatch(startLoginEmailPassword(email, password))
  }

  const handleLoginGoogle = () => {
    dispatch(startGoogleLogin())
  }

  const { email, password } = formValues
  return (
    <>
      <h3 className="login__title">Login</h3>
      <form className="auth__form" onSubmit={handleLogin}>
        <input
          type="text"
          placeholder="email"
          name="email"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="password"
          name="password"
          autoComplete="off"
          value={password}
          onChange={handleInputChange}
        />
        <button className="pointer" type="submit" disabled={loading}>
          Ingresar
        </button>
        <hr />

        <div className="login__social-network">
          <p>Login with social network</p>
          <div
            className="google-btn pointer mt-1 mb-1"
            onClick={handleLoginGoogle}>
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src={googleLogo}
                alt="google button"
              />
            </div>
            <p className="btn-text">
              <b>Sign in with google</b>
            </p>
          </div>
        </div>

        <Link to="/auth/register">Create new account</Link>
      </form>
    </>
  )
}
