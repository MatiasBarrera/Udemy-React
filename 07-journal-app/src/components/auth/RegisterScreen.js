import React from 'react'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import isEmail from 'validator/lib/isEmail'
import { useDispatch, useSelector } from 'react-redux'
import { removeError, setError } from '../../actions/ui'
import { startRegisterWithEmailPasswordName } from '../../actions/auth'

export const RegisterScreen = () => {
  const dispatch = useDispatch()
  const { msgError } = useSelector(state => state.ui)

  const [formValues, handleInputChange] = useForm({
    name: 'matias',
    email: 'matias@gmail.com',
    password: '123456',
    password2: '123456'
  })

  const { name, email, password, password2 } = formValues

  const handleRegister = e => {
    e.preventDefault()

    if (isFormValid()) {
      dispatch(startRegisterWithEmailPasswordName(name, email, password))
    }
  }

  const isFormValid = () => {
    if (name.trim().length === 0) {
      dispatch(setError('Name is required'))
      return false
    }
    if (email.trim().length === 0 && !isEmail(email)) {
      dispatch(setError('Email is not valid'))
      return false
    }
    if (password !== password2 || password.length < 5) {
      dispatch(
        setError(
          'Password should be at least 6 characters and match each other '
        )
      )
      return false
    }
    dispatch(removeError())
    return true
  }

  return (
    <>
      <h3 className="login__title">Register</h3>
      <form className="auth__form" onSubmit={handleRegister}>
        {msgError && <div className="auth__alert-error">{msgError}</div>}
        <input
          type="text"
          placeholder="name"
          name="name"
          autoComplete="off"
          value={name}
          onChange={handleInputChange}
        />
        <input
          type="text"
          placeholder="email"
          name="email"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="password"
          name="password"
          autoComplete="off"
          value={password}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="confirm"
          name="password2"
          autoComplete="off"
          value={password2}
          onChange={handleInputChange}
        />
        <button className="pointer" type="submit">
          Register
        </button>
        <hr />

        <Link to="/auth/login">Already registered?</Link>
      </form>
    </>
  )
}
