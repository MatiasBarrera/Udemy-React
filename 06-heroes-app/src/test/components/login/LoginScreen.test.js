import React from 'react'

import { mount } from 'enzyme'
import { LoginScreen } from '../../../components/login/LoginScreen'
import { Router } from 'react-router'
import { AuthContext } from '../../../auth/AuthContext'
import { types } from '../../../types/types'

describe('Pruebas en LoginScreen', () => {
  const historyMock = {
    push: jest.fn(),
    replace: jest.fn(),
    location: '',
    listen: jest.fn(),
    createHref: jest.fn()
  }

  const contextValue = {
    dispatch: jest.fn(),
    user: {
      logged: true,
      name: 'Matias'
    }
  }

  const wrapper = mount(
    <AuthContext.Provider value={contextValue}>
      <Router history={historyMock}>
        <LoginScreen />
      </Router>
    </AuthContext.Provider>
  )

  test('Debe de mostrar correctamente', () => {
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('h1').text().trim()).toBe('Welcome')
  })

  test('Debe de realizar el dispatch y la navegacion', () => {
    wrapper.find('button').simulate('click')
    expect(contextValue.dispatch).toHaveBeenCalledWith({
      type: types.login,
      payload: {
        name: 'Matias'
      }
    })

    expect(historyMock.replace).toHaveBeenCalled()
  })

  afterEach(() => {
    jest.clearAllMocks()
  })
})
