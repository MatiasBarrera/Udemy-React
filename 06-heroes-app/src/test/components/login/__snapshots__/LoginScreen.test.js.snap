// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en LoginScreen Debe de mostrar correctamente 1`] = `
<div
  className="login-container"
>
  <div
    className="login-card-container"
  >
    <div
      className="login-card-header"
    >
      <h1>
        Welcome
      </h1>
    </div>
    <div
      className="login-card-body"
    >
      <button
        onClick={[Function]}
      >
        Login
      </button>
    </div>
  </div>
</div>
`;
